File: 2095-minimum-number-of-swaps-to-make-the-string-balanced/minimum-number-of-swaps-to-make-the-string-balanced.py
class Solution:
    def minSwaps(self, s: str) -> int:
        left, right = 0, 0

        for _s in s:
            if _s=='[':
                right+=1
            elif  _s==']' and right > 0:
                right-=1
            else:
                left+=1
        #print(left)
        #return (left +1)//2
        return ceil(left/2)

File: 2095-minimum-number-of-swaps-to-make-the-string-balanced/minimum-number-of-swaps-to-make-the-string-balanced.java
class Solution {
    public int minSwaps(String s) {

        // 0  1  2  3  4  5
        // ]  ]  ]  [  [  [
        // [. ]. ] [.  [ ]
        //

        // op : len(op)
        // swap: len(cl), cl > op

        // since string consists of n/2 and n/2 you can return either

        Stack<Integer> st = new Stack<>();
        Set<Integer> close_to_remove = new HashSet<>();

        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);
            if (ch == '[') {
                st.add(i);
            }
            else {
                if(!st.empty()){
                    st.pop();
                }
                else {
                    close_to_remove.add(i);
                }
            }
            
        }
        // # each swap fixes two, +1 for odd number sneeds 1 more
        return (close_to_remove.size() + 1)/ 2;
        
    }
}